src/reducers/__tests__/console.test.jsx

```diff
@@ -0,0 +1,163 @@
+import forEach from 'lodash-es/forEach';
+import map from 'lodash-es/map';
+import reduce from 'lodash-es/reduce';
+
+import reducer from '../console';
+
+import {
+  consoleErrorProduced,
+  consoleLogBatchProduced,
+  consoleValueProduced,
+  evaluateConsoleEntry,
+  nextConsoleHistory,
+  previousConsoleHistory,
+  consoleInputChanged,
+} from '../../actions';
+
+test('evaluateConsoleEntry adds entry to history', () => {
+  const expression = '1 + 1';
```

Comment:
These consts could be a Rosie Factory object, where key is '123' and maybe you can override with that expression and value when appropriate

Reply:
Hmm, so I generally think of the main goals of using a factory like Rosie to be to reduce duplication and to centralize knowledge about object structure in one place. Do you think either of those goals would be served by bringing a factory into the picture here? What kind of objects would the factory actually produce? Where in the application code would objects with the same shape appear?

I would characterize what we’re doing here as creating test sentinels—values that are basically meaningless except that we want to give them as input to the code under test, and then check to see if they are in the right place in the output. This is basically the same thing that factories do, but factories always produce objects, and specifically they product objects with a shape that you would find somewhere else in application code (e.g. we are using factories to produce objects with the same shape that the GitHub and Firebase SDKs produce, etc.)

So I guess my question is, what shape object are you expecting to see here? What would the factories actually produce?

Reply:
@jwang1919 so the most obvious answer I could think of for what a factory would generate here was the action itself—essentially wrapping the action creator but having sensible defaults/sequences for all the args that get passed in. Take a look at [this commit](https://github.com/popcodeorg/popcode/pull/1687/commits/f11a2380365a6bb0dd8f3858b27f463770fb1ca0) and see if you think it’s an improvement; if so, I will apply the pattern to all the action creators used in that test.

Reply:
The intention was to remove the repetitive lines of code where `key`, `value`, `expression` and `compiledProject` were declared. The secondary intention was to avoid hard-coding expected values in the assertions by using the factory created objects instead.

I only thought Rosie was being used as a means to generate test objects. I do agree it makes more sense to have Rosie generate more sensible objects that can be seen in the application.

TLDR; I'm okay skipping this suggestion.

Reply:
Gotcha. What do you think of the [experiment with using factories to generate actions in tests](https://github.com/popcodeorg/popcode/pull/1687/commits/f11a2380365a6bb0dd8f3858b27f463770fb1ca0)? I am on the fence—I like removing the burden of generating sentinel values from the tests themselves, but I worry that it creates extra test boilerplate without a really compelling improvement to abstraction/encapsulation. Also, I think it’s probably better to have tests coupled to the interface of action creators (by calling them directly) than to the output of action creators (by extracting values from the action payloads to make assertions about).

Reply:
Yeah I would skip it, I think it is extra syntactic sugar that would confuse new developers.

Reply:
OK!